### ISSUE TITLE
BUG: IntrTriangle3Triangle3.Test gives wrong result on diabolical triangles

### ISSUE THREAD

**Issue by Garrett-R** [gradientspace/geometry3Sharp#85]
This method doesn't work if one or more vertices of one of the triangles have the same coordinates.

Consider this example:

```C#
using System;
using g3;

namespace ConsoleApp1 {
    class Program {
        static void Main(string[] args) {
            Triangle3d t0 = new Triangle3d(
                new Vector3d(0, 0, 0),
                new Vector3d(0, 0, 0),
                new Vector3d(0, 0, 0)
            );
            Triangle3d t1 = new Triangle3d(
                new Vector3d(0, 0, 0),
                new Vector3d(0, 0, 0),
                new Vector3d(0, 1, 0)
            );
            Triangle3d t2 = new Triangle3d(
                new Vector3d(0, 0, 1),
                new Vector3d(0, 1, 1),
                new Vector3d(1, 1, 1)
            );

            var intersect = new IntrTriangle3Triangle3(t0, t1);
            Console.WriteLine($"t0 and t1 intersects: {intersect.Test()}");
            var intersect2 = new IntrTriangle3Triangle3(t1, t2);
            Console.WriteLine($"t1 and t2 intersects: {intersect.Test()}");

            Console.ReadLine();
        }
    }
}
```

This will print:
```
true
true
```

(not being pedantic, I actually had this in a real project, although I fixed it by just filtering out useless, zero-area triangles)

**Comment by rms80** [comment]
Find() might be slightly more informative, but yes, when your triangle is actually a point or a line, this code is not going to work at all, as one if the first things it does is try to compute the cross-product of the two edges. 

(t0 and t1 do intersect, though ;)

I don't really consider this a bug, more of a Garbage-in, Garbage-out situation. Catching all these cases and switching to line-triangle intersection, that sort of thing, is possible but then you inherit that overhead in the not-degenerate case. Possibly it could throw an exception if the cross-product is degenerate, but then everyone calling this needs to always be catching exceptions, another overhead. 

Feel free to submit a PR if you have an idea of how this should be handled =)

**Comment by Garrett-R** [comment]
Yeah, I was also wondering about the trade-off between diabolical input and performance.  I'm relatively new to comp. geometry, so not sure how frequent it is to come across degenerate triangles.  In my case, the software I'm writing a plugin for gave some of these when triangulating a mesh which is kind of surprising.

In terms of the performance hit to the non-degenerate cases, I suppose the line to check for this edge case could go be inserted above [here](https://github.com/gradientspace/geometry3Sharp/blob/master/intersection/IntrTriangle3Triangle3.cs#L212) since with degenerate triangles, you will necessarily hit that line.  

This means the performance hit will be felt by only coplanar triangles. the cost will be an if statement with 6 Vector3D equalities:

```C#
if (triangle0.V0 == triangle0.V1 || triangle0.V1 == triangle0.V2 || triangle0.V2 == triangle0.V0 || triangle1.V0 == triangle1.V1 || triangle1.V1 == triangle1.V2 || triangle1.V2 == triangle1.V0)
```

and of course will add some complexity to the code.  Might not look pretty inside that `if`.  I think a new class would also be required (`IntrLine3Triangle3`).

Anyway, can't promise that I'll be able to submit a PR for this.  Do you think it's worth the complexity/performance cost?  If I'm the only person that's run into this, maybe not...

**Comment by Garrett-R** [comment]
Oh wait, I think it's more simple actually.  We could do the above `if` statement, and then fall-back to using `Find()` if we have this rare edge case.  I assume `Find()` is slower otherwise `Test` would just call that, but I think that's fine because performance for the diabolical input case is not really too important.  This would keep the code clean.

There might be a bit of complication trying to pull apart the methods with side-effects (so that this works with the static method `Intersects`).  Anyway, yeah let me know if you think it's worth it and I can take a look.

**Comment by Garrett-R** [comment]
Oh, very interesting.  Yeah, that all makes sense.  Thanks for the explanation!
